src/Middleware/ResponseCompression/src/BodyWrapperStream.cs

```diff
@@ -15,35 +16,37 @@ namespace Microsoft.AspNetCore.ResponseCompression
     /// <summary>
     /// Stream wrapper that create specific compression stream only if necessary.
     /// </summary>
-    internal class BodyWrapperStream : Stream, IHttpBufferingFeature, IHttpSendFileFeature, IHttpResponseStartFeature, IHttpsCompressionFeature
+    internal class BodyWrapperStream : Stream, IHttpResponseBodyFeature, IHttpsCompressionFeature
     {
         private readonly HttpContext _context;
-        private readonly Stream _bodyOriginalStream;
         private readonly IResponseCompressionProvider _provider;
-        private readonly IHttpBufferingFeature _innerBufferFeature;
-        private readonly IHttpSendFileFeature _innerSendFileFeature;
-        private readonly IHttpResponseStartFeature _innerStartFeature;
+        private readonly IHttpResponseBodyFeature _innerBodyFeature;
 
         private ICompressionProvider _compressionProvider = null;
         private bool _compressionChecked = false;
         private Stream _compressionStream = null;
+        private PipeWriter _pipeAdapter = null;
         private bool _providerCreated = false;
         private bool _autoFlush = false;
 
-        internal BodyWrapperStream(HttpContext context, Stream bodyOriginalStream, IResponseCompressionProvider provider,
-            IHttpBufferingFeature innerBufferFeature, IHttpSendFileFeature innerSendFileFeature, IHttpResponseStartFeature innerStartFeature)
+        internal BodyWrapperStream(HttpContext context, IResponseCompressionProvider provider,
+            IHttpResponseBodyFeature innerBodyFeature)
         {
             _context = context;
-            _bodyOriginalStream = bodyOriginalStream;
             _provider = provider;
-            _innerBufferFeature = innerBufferFeature;
-            _innerSendFileFeature = innerSendFileFeature;
-            _innerStartFeature = innerStartFeature;
+            _innerBodyFeature = innerBodyFeature;
         }
 
-        internal ValueTask FinishCompressionAsync()
+        internal async Task FinishCompressionAsync()
         {
-            return _compressionStream?.DisposeAsync() ?? new ValueTask();
+            if (_pipeAdapter != null)
```

Comment:
This should noop if called twice.

Reply:
CompleteAsync and DisposeAsync already do that internally.

Reply:
We should do a check since we always call it at the end of the middleware pipeline.